include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -1673,117 +1549,91 @@ struct __parallel_sort_submitter<__internal::__optional_kernel_name<_LeafSortNam
                                  __internal::__optional_kernel_name<_GlobalSortName...>,
                                  __internal::__optional_kernel_name<_CopyBackName...>>
 {
-    template <typename _ExecutionPolicy, typename _Range, typename _Merge, typename _Compare>
+    template <typename _ExecutionPolicy, typename _Range, typename _Compare>
     auto
-    operator()(_ExecutionPolicy&& __exec, _Range&& __rng, _Merge __merge, _Compare __comp) const
+    operator()(_ExecutionPolicy&& __exec, _Range&& __rng, _Compare __comp) const
     {
         using _Policy = typename ::std::decay<_ExecutionPolicy>::type;
         using _Tp = oneapi::dpl::__internal::__value_t<_Range>;
         using _Size = oneapi::dpl::__internal::__difference_t<_Range>;
 
-        _Size __n = __rng.size();
+        const ::std::size_t __n = __rng.size();
         assert(__n > 1);
 
-        _PRINT_INFO_IN_DEBUG_MODE(__exec);
-
-        // __leaf: size of a block to sort using algorithm suitable for small sequences
-        // __optimal_chunk: best size of a block to merge duiring a step of the merge sort algorithm
-        // The coefficients were found experimentally
-        _Size __leaf = 4;
-        _Size __optimal_chunk = 4;
-        if (__exec.queue().get_device().is_cpu())
-        {
-            __leaf = 16;
-            __optimal_chunk = 32;
-        }
-        // Assume powers of 2
-        assert((__leaf & (__leaf - 1)) == 0);
-        assert((__optimal_chunk & (__optimal_chunk - 1)) == 0);
-
-        const _Size __leaf_steps = ((__n - 1) / __leaf) + 1;
+        const ::std::uint16_t __chunk = 4;
+        const _Size __steps = oneapi::dpl::__internal::__dpl_ceiling_div(__n, __chunk);
 
         // 1. Perform sorting of the leaves of the merge sort tree
-        sycl::event __event1 = __exec.queue().submit([&](sycl::handler& __cgh) {
+        sycl::event __event1 = __exec.queue().submit([&, __comp](sycl::handler& __cgh) {
             oneapi::dpl::__ranges::__require_access(__cgh, __rng);
-            __cgh.parallel_for<_LeafSortName...>(sycl::range</*dim=*/1>(__leaf_steps),
-                                                 [=](sycl::item</*dim=*/1> __item_id) {
-                                                     const _Size __idx = __item_id.get_linear_id() * __leaf;
-                                                     const _Size __start = __idx;
-                                                     const _Size __end = sycl::min(__start + __leaf, __n);
-                                                     __leaf_sort_kernel()(__rng, __start, __end, __comp);
-                                                 });
-        });
+            __cgh.parallel_for<_LeafSortName...>(sycl::range</*dim=*/1>(__steps), [=](sycl::item</*dim=*/1> __item_id)
+            {
+                const ::std::uint32_t i_elem = __item_id.get_linear_id() * __chunk;
```

Comment:
1) Actually, 64 bit index variable (::std::size_t) leads to performance degradation down to 20% (PVC) ....
2) ::std::uint32_t "covers" 0...4B elements that is very bigger that we can launch practically.

Reply:
What limitation stops 4B elements from being possible?  
For integers, 4 Billion is 16 GB, which is still possible to handle.  Is there a maximum total work items?

Reply:
Theoretically, yes... it can be more than 4B... 
But, what should we do to eliminate the performance degradation down to 20%  in case of std::size_t ?
To have a couple of kernels? I don't see another option....

Reply:
To me, the best option is a second kernel which handles the cases for very large N (same code, just templated type for the index), and is launched with a runtime check on the size.  It is definitely frustrating that this would be required.

Its surprising that this is such a high performance penalty.  I wonder if we are paying a similar penalty in other algorithms.

Reply:
Fixed